File: Scala/src/com/yourtion/Pattern16/PartialExamples.scala
package com.yourtion.Pattern16

/**
  * Created by Yourtion on 05/04/2017.
  */
object PartialExamples {

  def addTwoInts(intOne: Int, intTwo: Int) = intOne + intTwo

  def taxForState(amount: Double, state: Symbol) = state match {
    // Simple tax logic, for example only!
    case ('NY) => amount * 0.0645
    case ('PA) => amount * 0.045
    // Rest of states...
  }

  def run(): Unit = {
    println("PartialExamples :")
    val addFortyTwo = addTwoInts(42, _: Int)
    println("100 addFortyTwo: " + addFortyTwo(100))

    val nyTax = taxForState(_: Double, 'NY)
    println("nyTax 100 : " + nyTax(100))

    val paTax = taxForState(_: Double, 'PA)
    println("paTax 100 : " + paTax(100))

  }

}

File: Clojure/src/com/yourtion/Pattern16/partial_examples.clj
(ns com.yourtion.Pattern16.partial-examples)

(defn add-two-ints [int-one int-two] (+ int-one int-two))

(defn tax-for-state [state amount]
  (cond
    (= :ny state) (* amount 0.0645)
    (= :pa state) (* amount 0.045)))

(defn run []
  (println "PartialExamples :")
  (def add-fourty-two (partial add-two-ints 42))
  (println "100 addFortyTwo: " (add-fourty-two 100))
  (def ny-tax (partial tax-for-state :ny))
  (println "nyTax 100 : " (ny-tax 100))
  (def pa-tax (partial tax-for-state :pa))
  (println "paTax 100 : " (pa-tax 100))
  (println ""))
